src/hasProp.test.ts

```diff
@@ -0,0 +1,138 @@
+import { hasProp } from "./hasProp";
+import { pipe } from "./pipe";
+
+describe("data-first", () => {
+  it("should return true when a prop is present on the object and is defined", () => {
+    const fooSymbol = Symbol("foo");
+    const obj = { a: 1, 1: "hi", [fooSymbol]: "bar" };
+    expect(hasProp(obj, "a")).toBe(true);
+    expect(hasProp(obj, 1)).toBe(true);
+    expect(hasProp(obj, fooSymbol)).toBe(true);
+  });
+
+  it("should return false when a prop is not present on the object", () => {
+    const fooSymbol = Symbol("foo");
+    const obj = { a: 1 };
+    // @ts-expect-error this property does not exist on the object
+    expect(hasProp(obj, "b")).toBe(false);
+    // @ts-expect-error this property does not exist on the object
+    expect(hasProp(obj, 1)).toBe(false);
+    // @ts-expect-error this property does not exist on the object
+    expect(hasProp(obj, fooSymbol)).toBe(false);
+  });
+
+  it("should return false when a prop is present on the object but is undefined", () => {
+    const obj = { b: undefined };
+    expect(hasProp(obj, "b")).toBe(false);
+  });
```

Comment:
This is incorrect and relies on a more permissive type definition. When the typescript project uses `exactOptionalPropertyTypes` an `undefined` value is strictly different than an optional prop not being present. This is also how it works during runtime.

```ts
console.log('a' in { a: undefined }); // "true"
console.log('a' in {}); // "false"
```

Reply:
This behavior was an intentional trade-off. I can play around a bit more since it's been awhile, but I remember something at the type level couldn't reliably distinguish between the two.

Reply:
Check out type-fest types for some of your utility types under _objectProps, they most likely can solve most of those issues.

hasProp should work as closely as possible to `"prop" in <obj>` otherwise some users would either misunderstand it's semantics and run into problems in runtime, or need to build their own function for the behaviour they need.

`hasPropSatisfying(obj, prop, isDefined.strict)` can provide the functionality you define for `hashProp` here, but the opposite is not possible.

Reply:
I'll check it out and see what I can do. This is the issue I reference elsewhere in the PR that lead me to this decision. https://github.com/microsoft/TypeScript/issues/31025